src/__tests__/extensions/toolbar.test.ts

```diff
@@ -0,0 +1,199 @@
+            access_token: 'access token',
+            [key]: encodeURIComponent(_isString(hashState) ? hashState : JSON.stringify(hashState)),
+            expires_in: 3600,
+        })
+
+        const withHash = (hashParams: Record<string, any>) => {
+            return Object.keys(hashParams)
+                .map((k) => `${k}=${hashParams[k]}`)
+                .join('&')
+        }
+
+        const aLocation = (hash?: string) => {
+            if (_isUndefined(hash)) {
+                hash = withHash(withHashParamsFrom())
+            }
+
+            return {
+                hash: `#${hash}`,
+                pathname: 'pathname',
+                search: '?search',
+            }
+        }
+
+        beforeEach(() => {
+            localStorage.getItem.mockImplementation(() => {})
+
+            jest.spyOn(toolbar, 'loadToolbar')
+        })
+
+        it('should initialize the toolbar when the hash state contains action "ph_authorize"', () => {
+            // the default hash state in the test setup contains the action "ph_authorize"
+            toolbar.maybeLoadToolbar(aLocation(), localStorage, history)
+
+            expect(toolbar.loadToolbar).toHaveBeenCalledWith({
+                ...toolbarParams,
+                ...defaultHashState,
+                source: 'url',
+            })
+        })
+
+        it('should initialize the toolbar when there are editor params in the session', () => {
+            // if the hash state does not contain ph_authorize then look in storage
+            localStorage.getItem.mockImplementation(() => JSON.stringify(toolbarParams))
+
+            const hashState = { ...defaultHashState, action: undefined }
+            toolbar.maybeLoadToolbar(aLocation(withHash(withHashParamsFrom(hashState))), localStorage, history)
+
+            expect(toolbar.loadToolbar).toHaveBeenCalledWith({
+                ...toolbarParams,
+                source: 'localstorage',
+            })
+        })
+
+        it('should NOT initialize the toolbar when the activation query param does not exist', () => {
+            expect(toolbar.maybeLoadToolbar(aLocation(''), localStorage, history)).toEqual(false)
+
+            expect(toolbar.loadToolbar).not.toHaveBeenCalled()
+        })
+
+        it('should return false when parsing invalid JSON from fragment state', () => {
+            expect(
+                toolbar.maybeLoadToolbar(aLocation(withHash(withHashParamsFrom('literally'))), localStorage, history)
+            ).toEqual(false)
+            expect(toolbar.loadToolbar).not.toHaveBeenCalled()
+        })
+
+        it('should work if calling toolbar params `__posthog`', () => {
+            toolbar.maybeLoadToolbar(
+                aLocation(withHash(withHashParamsFrom(defaultHashState, '__posthog'))),
+                localStorage,
+                history
+            )
+            expect(toolbar.loadToolbar).toHaveBeenCalledWith({ ...toolbarParams, ...defaultHashState, source: 'url' })
+        })
```

Comment:
this test changed as well as being converted, it wasn't actually testing what it said it was